src/dune_pkg/opam_solver.ml

```diff
@@ -195,19 +198,85 @@ module Context_for_dune = struct
     { available; resolved }
   ;;
 
+  (* Move all packages with the avoid-version flag beyond any package
+     without that flag, otherwise preserving the order. This is needed
+     so that development versions of packages (such as alpha releases
+     of the compiler) are not favoured over older, stable releases of
+     the package. This is a hack to work around the fact that
+     opam-0install-solver does not implement the avoid-versions
+     flag. *)
+  let move_avoid_versions_packages_to_end packages =
```

Comment:
I don't think this is needed? This is already handled by the new comparison function.

Reply:
It is still needed, as we need to make sure the most recent version of the `ocaml` package that depends on a non-`avoid-version` compiler package is located in front of versions of `ocaml` that have only packages that are marked `avoid-version`.

E.g. the current state of the [`ocaml.5.3.0`](https://github.com/ocaml/opam-repository/blob/0b240d2960133fd3d8aa8f008d7aa79534caa3b9/packages/ocaml/ocaml.5.3.0/opam) depends on `ocaml-base-compiler.5.3.0` (doesn't exist) or `ocaml-variants {>= "5.3.0~" & < "5.3.1~"}` which can be fulfilled by [`ocaml-variants.5.3.0+trunk`](https://github.com/ocaml/opam-repository/blob/0b240d2960133fd3d8aa8f008d7aa79534caa3b9/packages/ocaml-variants/ocaml-variants.5.3.0%2Btrunk/opam).

Thus without this function the solver would attempt to pick `ocaml.5.3.0`; see that there are no candidates for `ocaml-base-compiler` and skip; see that there is one candidate for `ocaml-variants` (we have sorted it to the back due to `avoid-version` but the candidates list only contains one entry anyway) and pick that one.

This is a consistent pattern in opam-repository (that there is a new `ocaml` package before the final release of `ocaml-base-compiler` exists) that in many cases Dune would always pick `ocaml-variants.<latest-development-release>+trunk` that this hack might be worth it.

Reply:
I'm going to break the logic for restricting the ocaml version to the latest stable ocaml-base-compiler version into its own PR for clarity, then close this PR.